### ISSUE TITLE
BenchmarkResult schema: reconcile, simplify `error` vs `validation`

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#832]
Follow-up from https://github.com/conbench/conbench/issues/825#issuecomment-1459903081 and https://github.com/conbench/conbench/issues/813.

`error` and `validation` and `stats` are not orthogonal to each other. All three can as of today be used to represent error information. How should users decide?

I described [elsewhere](https://gehrcke.de/2022/11/interface-design-a-craft-and-an-art/) that

> Maximizing simplicity requires identification of the fundamental degrees of freedom (that then need to be exposed). Removing variables. Identifying invariants, and working with them. This kind of thinking can make an interface justifiably beautiful.

I think this here is a great example. The goal is not to remove flexibility, but to remove ambiguity.

The first goal of this ticket is to have a nice piece of documentation that describes users what to use these fields for, and how. 

The second goal is to have updated schemas / input data validation, which is helpfully in sync with the new documentation/specification.

Quoting @jonkeane about the current state:

> The major (only?) difference between error and validation is that if there’s contents in error, the data from the benchmarks are excluded from results, history, etc.  [...] But for validation that’s informational only.

There are historical reasons for introducing `validation`, but we should question those.

In unit testing there often is `passed, failed, skipped`. But when the test _has_ an outcome, it's binary.

I think it makes sense for a benchmark result to be in either of precisely two states: passed (data is being used), failed (data is not being used).

If users believe they need something inbetween, then my answer is that the definition of "passed" vs. "failed" needs to be worked on.

Simplistic, powerful view:

- when a benchmark result is _failed_ then there must be error information
- when a benchmark result is _passed_ then there must not be error information.

The word "validation" has strong semantics which overlaps with the concept of error. I think it only dilutes the view above.

Maybe sometimes in addition to "passed", for some unknown meta use case that has nothing to do with benchmarking, people want to use the Conbench system as a vehicle to cheaply persist some useful data which happens to have been obtained during benchmarking, as a side-effect. We _could_ provide a free text / JSON field where a benchmarking environment can annotate a benchmark result with some additional info; the use case and semantic meaning should be completely neutral.

My intuition is that we should remove 'validation'.

**Comment by jgehrcke** [comment]
> We could provide a free text / JSON field where a benchmarking environment can annotate a benchmark result with some additional info; the use case and semantic meaning should be completely neutral.

It seems like we have that already: `optional_benchmark_info`, currently documented as 

> Optional information about Benchmark results (e.g., telemetry links, logs links). These are unique to each benchmark that is run, but are information that aren't reasonably expected to impact benchmark performance. Helpful for adding debugging or additional links and context for a benchmark (free-form JSON)

We also have `info` (required):

> Additional information about the context the benchmark was run in that is not expected to have an impact on benchmark performance (e.g. benchmark language version, compiler version). This information is expected to be the same across a number of benchmarks. (free-form JSON)

**Comment by alistaire47** [comment]
Yeah, I think a validation error is just a specific type of error. It's useful for debugging to put in `error` what stage it failed at (setup, running, validation, etc.), but it's a failure regardless.

**Comment by jonkeane** [comment]
Yup agreed with lots of what y'all are saying here. 

> The major (only?) difference between error and validation is that if there’s contents in error, the data from the benchmarks are excluded from results, history, etc. [...] But for validation that’s informational only.

This is definitely an accurate state of the world today. 

> I think it makes sense for a benchmark result to be in either of precisely two states: passed (data is being used), failed (data is not being used).
> 
> If users believe they need something inbetween, then my answer is that the definition of "passed" vs. "failed" needs to be worked on.
> 
> Simplistic, powerful view:
> 
> * when a benchmark result is failed then there must be error information
> * when a benchmark result is passed then there must not be error information.
> The word "validation" has strong semantics which overlaps with the concept of error. I think it only dilutes the view above.
> 
> ... 
> 
> My intuition is that we should remove 'validation'.

Agreed. I think ultimately (my) adding of `validation` was a mistake here, and we should endeavor to correct that when we have the opportunity to. It would be much better to require `error` to be present if it's a true error (i.e. shouldn't be considered and used for data; e.g. a validation error, a setup error, a query crash error, etc all of which are errors). 

If a conbench user | instance wants to treat wrong-data as a non-error state (though we discourage this!) they absolutely can, so long as they do not populate the validation in the `error` within the field, but rather in another filed (`optional_benchmark_info` is a great candidate for that IMO, which y'all have already talked about). This then becomes something of an adaptor | connector | runner issue, but that is much closer to the actual implementation using this anyway.

This gives us our strong point of view about errors | validation (if it doesn't validate, something is seriously wroong), but allows us to support folks who disagree with that (of which I know one of one major conbench customer who does for right now). 

> Maybe sometimes in addition to "passed", for some unknown meta use case that has nothing to do with benchmarking, people want to use the Conbench system as a vehicle to cheaply persist some useful data which happens to have been obtained during benchmarking, as a side-effect. We could provide a free text / JSON field where a benchmarking environment can annotate a benchmark result with some additional info; the use case and semantic meaning should be completely neutral. 
> + the message below about `optional_benchmark_info`

Great finds on all of this. We do know some cases where folks are using conbench to be a repository of info like this (almost as an extended testing framework). While that's definitely _not_ what conbench is built to do, or our major use target, it's ok for folks to do that in some cases (especially since the features that help do that actually also help with debugability and actionability for more solely benchmarking only systems — e.g. getting a stack trace of what went wrong is super important for benchmark authors | maintainers to see what happened if something failed suddenly; it _also_ happens to help in this use of conbench.) We should be careful + hesitant about building too much stuff into conbench that facilitates this kind of use — but where there is overlap with the pure benchmarking uses cases, that's ok!

**Comment by jgehrcke** [comment]
Btw, thanks for the beautiful reply above, Jon. I'd like to comment on this:

>  We should be careful + hesitant about building too much stuff into conbench that facilitates this kind of use

Yes! This is a really valuable line of thinking. I think I'd love to link this to https://github.com/conbench/conbench/issues/1251.

**Comment by alistaire47** [comment]
I've taken a look through user repos and I couldn't find any remaining usages, so at this point I believe we're fine to deprecate `validation`.
